File: core/src/main/java/de/qabel/core/repository/sqlite/SqliteDropStateRepository.java
package de.qabel.core.repository.sqlite;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import de.qabel.core.repository.DropStateRepository;
import de.qabel.core.repository.exception.EntityNotFoundException;
import de.qabel.core.repository.exception.PersistenceException;

public class SqliteDropStateRepository implements DropStateRepository {
    public static final String TABLE_NAME = "drop_state";
    private ClientDatabase database;

    public SqliteDropStateRepository(ClientDatabase database) {
        this.database = database;
    }

    @Override
    public String getDropState(String drop) throws EntityNotFoundException, PersistenceException {
        try (PreparedStatement statement = database.prepare(
            "SELECT `last_request_stamp` FROM " + TABLE_NAME + " WHERE `drop` = ? LIMIT 1"
        )) {
            statement.setString(1, drop);
            try (ResultSet resultSet = statement.executeQuery()) {
                if (!resultSet.next()) {
                    throw new EntityNotFoundException("no state found for drop '" + drop + "'");
                }

                return resultSet.getString(1);
            }
        } catch (SQLException e) {
            throw new PersistenceException("failed to load drop state: " + e.getMessage(), e);
        }
    }

    @Override
    public void setDropState(String drop, String state) throws PersistenceException {
        try (PreparedStatement statement = database.prepare(
            "INSERT INTO " + TABLE_NAME + " (`drop`, `last_request_stamp`) VALUES (?, ?)"
        )) {
            statement.setString(1, drop);
            statement.setString(2, state);
            statement.execute();
        } catch (SQLException e) {
            throw new PersistenceException("failed to save drop " + e.getMessage(), e);
        }
    }
}

File: core/src/main/java/de/qabel/core/repository/sqlite/SqliteDropStateRepository.kt
package de.qabel.core.repository.sqlite

import de.qabel.core.drop.DropURL
import de.qabel.core.repository.DropStateRepository
import de.qabel.core.repository.EntityManager
import de.qabel.core.repository.entities.DropState
import de.qabel.core.repository.exception.EntityNotFoundException
import de.qabel.core.repository.framework.BaseRepository
import de.qabel.core.repository.sqlite.hydrator.DropStateAdapter
import de.qabel.core.repository.sqlite.schemas.DropStateDB

class SqliteDropStateRepository(database: ClientDatabase,
                                entityManager: EntityManager) :
    BaseRepository<DropState>(DropStateDB, DropStateAdapter(), database, entityManager), DropStateRepository {

    override fun getDropState(drop: String) = findByDropId(drop).eTag

    private fun findByDropId(dropId: String): DropState =
        with(createEntityQuery()) {
            whereAndEquals(DropStateDB.DROP, dropId)
            return getSingleResult<DropState>(this)
        }

    override fun setDropState(drop: String, state: String): Unit =
        try {
            findByDropId(drop).let {
                it.eTag = state
                update(it)
            }
        } catch(ex: EntityNotFoundException) {
            DropState(drop, state).let { persist(it) }
        }

    override fun getDropState(dropUrl: DropURL): DropState =
        try {
            findByDropId(dropUrl.toString())
        } catch(ex: EntityNotFoundException) {
            DropState(dropUrl.toString())
        }

    override fun setDropState(dropState: DropState) {
        if (dropState.id == 0) persist(dropState) else update(dropState)
    }

}

